lib/stm32/f7/flash.c

```diff
@@ -0,0 +1,478 @@
+/** @addtogroup flash_file
+ *
+ */
+
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2017 Matthew Lai @m@matthewlai.ca>
+ * Copyright (C) 2010 Thomas Otto <tommi@viadmin.org>
+ * Copyright (C) 2010 Mark Butler <mbutler@physics.otago.ac.nz>
+ *
+ * This library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**@{*/
+
+#include <libopencm3/stm32/flash.h>
+
+/*---------------------------------------------------------------------------*/
+/** @brief Set the Program Parallelism Size
+
+Set the programming word width. Note carefully the power supply voltage
+restrictions under which the different word sizes may be used. See the
+programming manual for more information.
+@param[in] psize The programming word width one of: @ref flash_cr_program_width
+*/
+
+static inline void flash_set_program_size(uint32_t psize)
+{
+	FLASH_CR &= ~(FLASH_CR_PROGRAM_MASK << FLASH_CR_PROGRAM_SHIFT);
+	FLASH_CR |= psize << FLASH_CR_PROGRAM_SHIFT;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Issue Pipeline Stall
+
+Issue a pipeline stall to make sure all write operations completed.
+
+RM0385: After performing a data write operation and before polling the BSY bit
+to be cleared, the software can issue a DSB instruction to guarantee the 
+completion of a previous data write operation.
+
+*/
+
+static inline void flash_pipeline_stall(void)
+{
+	asm volatile("dsb":::"memory");
```

Comment:
We've fixed it by specifying c99 in the makefiles, but see https://github.com/libopencm3/libopencm3/pull/769/commits/aab09823475953e855ec497a951fe8269443dbb9 for more info.  (just use __asm__ all the time for portability)